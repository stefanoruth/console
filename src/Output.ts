import rl from 'readline'

const colorList = {
	reset: '\x1b[0m',
	cyan: '\x1b[36m',
	yellow: '\x1b[33m',
}

type Colors = keyof typeof colorList

export class Output {
	/**
	 * Write a line to the console.
	 */
	line(message: string, newLine: boolean = false, color: Colors = 'reset'): void {
		process.stdout.write(color + '%s' + colorList.reset, message)

		if (newLine) {
			this.newLine()
		}
	}

	/**
	 * Display a table on the console.
	 */
	table(...args: any[]): void {
		console.table(args)
	}

	/**
	 * Display a set of new lines.
	 */
	newLine(count: number = 1) {
		process.stdout.write('\n'.repeat(count))
	}

	/**
	 * Ask the user for a question.
	 */
	ask(question: string): Promise<string> {
		return new Promise(resolve => {
			const r = rl.createInterface({
				input: process.stdin,
				output: process.stdout,
			})

			r.question(question + '\n', (answer: string) => {
				r.close()
				resolve(answer)
			})
		})
	}

	/**
	 * Ask the user if they really wanna do it.
	 */
	confirm(question: string, defaultValue: boolean = true): Promise<boolean> {
		return new Promise(resolve => {
			const r = rl.createInterface({
				input: process.stdin,
				output: process.stdout,
			})

			r.question(question + '\n', (answer: string) => {
				r.close()
				resolve(answer === 'Y')
			})
		})
	}
}
