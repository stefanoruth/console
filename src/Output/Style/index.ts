export * from './TextStyle'
export * from './Color'
export * from './NullColor'
