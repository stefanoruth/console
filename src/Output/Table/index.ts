export * from './Table'
export * from './TableDivider'
export * from './TableHeader'
export * from './TableRow'
export * from './TableStyle'
