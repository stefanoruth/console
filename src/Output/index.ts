export * from './ErrorHandler'
export * from './ProgressBar'
export * from './Question'
export * from './Style'
export * from './Table'
export * from './Descriptor'
export * from './Output'
export * from './OutputFormatter'
export * from './Terminal'
export * from './Verbosity'
export * from './Writer'
