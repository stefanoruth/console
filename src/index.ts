export { Application } from './Application'
export { Input, Option, Signature, Argument } from './Input'
export { Output, Terminal, ProgressBar, Table } from './Output'
export { Command } from './Commands'
export * from './Events/EventCollection'
export * from './Exceptions'
