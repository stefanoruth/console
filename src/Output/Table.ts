import { Output } from './Output'

export class Table {
	constructor(protected output: Output) {}

	render() {
		//
	}
}
