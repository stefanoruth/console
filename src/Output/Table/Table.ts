import { Output } from '../Output'
import { TableStyle } from './TableStyle'
import { TableCell } from './TableCell'
import { TableRow } from './TableRow'
import { TableDivider } from './TableDivider'
import { TableHeader } from './TableHeader'

export class Table {
	/**
	 * Table headers.
	 */
	protected headers: string[] = []

	/**
	 * Table rows.
	 */
	protected rows: object[] = []

	/**
	 * Number of columns cache.
	 */
	protected numberOfColumns?: number

	/**
	 * Table styles.
	 */
	protected style: TableStyle = new TableStyle()

	/**
	 * User set column widths.
	 */
	protected columnWidths: number[] = []

	/**
	 * Construct the table builder.
	 */
	constructor(protected output: Output) {}

	/**
	 * Renders table to output.
	 *
	 * Example:
	 *
	 *     +---------------+-----------------------+------------------+
	 *     | ISBN          | Title                 | Author           |
	 *     +---------------+-----------------------+------------------+
	 *     | 99921-58-10-7 | Divine Comedy         | Dante Alighieri  |
	 *     | 9971-5-0210-0 | A Tale of Two Cities  | Charles Dickens  |
	 *     | 960-425-059-0 | The Lord of the Rings | J. R. R. Tolkien |
	 *     +---------------+-----------------------+------------------+
	 */
	render() {
		const renderTable: string[] = new TableHeader(this.headers, [10, 12, 8], this.style).render()

		this.output.writer.write(renderTable)
	}

	setHeaders(headers?: string[]) {
		if (headers) {
			this.headers = headers
		}

		return this
	}

	setRows(rows: object[]) {
		this.rows = rows

		return this
	}
}
