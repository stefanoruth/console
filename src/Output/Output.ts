import rl from 'readline'
import { Color, CliColor } from './CliColor'
import { ProgressBar } from './ProgressBar'
import { Table } from './Table'
import { Question } from './Question/Question'
import { Writer } from './Writer'

export class Output {
	constructor(public writer: Writer = new Writer(), protected color: CliColor = new CliColor()) {}

	/**
	 * Write a line to the console.
	 */
	line(message: string, newLine: boolean = true, color?: Color) {
		this.writer.write(this.color.apply(message, { text: color }), newLine)
	}

	/**
	 * Formats a command title.
	 */
	title(message: string) {
		//
	}

	/**
	 * Formats a section title.
	 */
	section(message: string) {
		//
	}

	/**
	 * Formats a list.
	 */
	listing(elements: any[]) {
		//
	}

	/**
	 * Formats a command comment.
	 */
	comment(message: string | string[]) {
		this.writer.block(message, undefined, undefined, '<fg=default;bg=default> // </>', false, false)
	}

	/**
	 * Formats a success result bar.
	 */
	success(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'green' }))
	}

	/**
	 * Formats an error result bar.
	 */
	error(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'white', bg: 'red' }))
		// this.block(message, 'ERROR', 'fg=white;bg=red', ' ', true)
	}

	/**
	 * Formats an warning result bar.
	 */
	warning(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'black', bg: 'yellow' }))
		// this.block(message, 'WARNING', 'fg=black;bg=yellow', ' ', true)
	}

	/**
	 * Formats a note admonition.
	 */
	note(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'yellow' }))
		// this.block(message, 'NOTE', 'fg=yellow', ' ! ')
	}

	/**
	 * Formats a caution admonition.
	 */
	caution(message: string) {
		this.writer.writeln(this.color.apply(message, { text: 'white', bg: 'red' }))
		// this.block(message, 'CAUTION', 'fg=white;bg=red', ' ! ', true)
	}

	/**
	 * Display a table on the console.
	 */
	table(rows: object[], columns?: object) {
		const table = new Table(this)
		table.setHeaders(columns)
		table.setRows(rows)

		table.render()
		this.newLine()
	}

	/**
	 * Display a set of new lines.
	 */
	newLine(count: number = 1) {
		this.writer.newLine(count)
	}

	/**
	 * Ask the user for a question.
	 */
	ask(question: string) {
		return new Question().ask(question)
	}

	askQuestion(question: Question) {
		//
	}

	/**
	 * Ask the user if they really wanna do it.
	 */
	confirm(question: string, defaultValue: boolean = true): Promise<boolean> {
		return new Promise(resolve => {
			const r = rl.createInterface({
				input: process.stdin,
				output: process.stdout,
			})

			r.question(question + '\n', (answer: string) => {
				r.close()
				resolve(answer.toUpperCase() === 'Y')
			})
		})
	}

	/**
	 * Start a new progress bar
	 */
	progressBar(max: number = 0): ProgressBar {
		const progressBar = new ProgressBar(this, max)

		progressBar.start()

		return progressBar
	}
}
