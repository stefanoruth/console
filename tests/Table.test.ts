import { Table, TableStyle, TableDivider, TableHeader, TableRow } from '../src/Output/Table'
import { Color, NullColor } from '../src/Output/Style'
import { Output } from '../src/Output'
import { Mock } from 'ts-mockery'

class TestTable extends Table {
	getWidth() {
		return this.columnWidths
	}

	getStyle() {
		return this.style
	}
}

function getMock() {
	const o = Mock.of<Output>({ raw: jest.fn() })

	return new TestTable(o, new NullColor())
}

describe('Table', () => {
	describe('Style', () => {
		test('Fill column to required space', () => {
			const s = new TableStyle()

			expect(s.pad('foo', 6)).toBe(' foo    ')
			expect(s.pad('foo', 0)).toBe(' foo ')
			expect(s.pad('foo', 6).length).toBe(8)
			expect(s.pad('foo', 0).length).toBe(5)
			expect(s.fillEmpty(0).length).toBe(2)
			expect(s.fillEmpty(2).length).toBe(4)
		})

		test('Can wrap rows in borders', () => {
			const s = new TableStyle()

			expect(s.wrapRow('foo')).toBe('|foo|')
		})

		test('Render a column divider if needed', () => {
			const s = new TableStyle()

			expect(s.columnDivider(false)).toBe('')
			expect(s.columnDivider(true)).toBe('|')
		})

		test('Can be initiated with different styles', () => {
			expect(new TableStyle({}).paddingChar).toBe(' ')
			expect(new TableStyle({ paddingChar: '#' }).paddingChar).toBe('#')
			expect(new TableStyle({ paddingChar: '$' }).paddingChar).toBe('$')
		})
	})

	describe('Divider', () => {
		test('Render a dividing line', () => {
			expect(new TableDivider('headerTop', [1, 2, 3], new TableStyle()).render()).toBe('+---+----+-----+')
		})
	})

	describe('Header', () => {
		test('Show a basic header', () => {
			expect(new TableHeader(['foo', 'bar'], [], new TableStyle(), new NullColor()).render()).toEqual([
				'+-----+-----+',
				'| foo | bar |',
				'+-----+-----+',
			])
		})

		test('Fill empty spaces', () => {
			expect(new TableHeader(['foo'], [3, 3], new TableStyle(), new NullColor()).render()).toEqual([
				'+-----+-----+',
				'| foo |     |',
				'+-----+-----+',
			])
		})
	})

	describe('Row', () => {
		test('Can wrap a row in a outer set', () => {
			expect(new TableRow({ foo: 'bar', a: 2 }, [], new TableStyle()).render()).toBe('| bar | 2 |')
			expect(new TableRow({ foo: 'bar' }, [3, 1], new TableStyle()).render()).toBe('| bar |   |')
		})
	})

	test('Table with no data', () => {
		const t = getMock()

		expect(t.render()).toEqual(['++', '||', '++', '++'])
	})

	test('Table with no headers', () => {
		const t = getMock()
		t.setRows([{ a: 'foo', b: 'bar' }])

		expect(t.render()).toEqual(['+-----+-----+', '| a   | b   |', '+-----+-----+', '| foo | bar |', '+-----+-----+'])
	})

	test('Table with headers', () => {
		const t = getMock()
		t.setRows([{ a: 'foo', b: 'bar' }])
		t.setHeaders(['A', 'B'])

		expect(t.render()).toEqual(['+-----+-----+', '| A   | B   |', '+-----+-----+', '| foo | bar |', '+-----+-----+'])
	})

	describe('Table Styles', () => {
		test('default style', () => {
			const t = getMock()

			t.setStyle('default')

			expect(t.getStyle().horizontalInsideBorderChar).toBe('-')
			expect(t.getStyle().verticalInsideBorderChar).toBe('|')
		})

		test('slim style', () => {
			const t = getMock()

			t.setStyle('slim')

			expect(t.getStyle().horizontalInsideBorderChar).toBe('─')
			expect(t.getStyle().verticalInsideBorderChar).toBe('│')
		})

		test('custom style', () => {
			const t = getMock()

			const s = new TableStyle({
				horizontalInsideBorderChar: 'a',
				verticalInsideBorderChar: 'b',
			})

			t.setStyle(s)

			expect(t.getStyle().horizontalInsideBorderChar).toBe('a')
			expect(t.getStyle().verticalInsideBorderChar).toBe('b')
		})
	})
})
