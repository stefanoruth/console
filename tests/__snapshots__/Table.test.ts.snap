// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Header Fill empty spaces 1`] = `
Array [
  "+-----+-----+",
  "| foo |     |",
  "+-----+-----+",
]
`;

exports[`Table Header Show a basic header 1`] = `
Array [
  "+-----+-----+",
  "| foo | bar |",
  "+-----+-----+",
]
`;

exports[`Table Table Styles custom style 1`] = `
Array [
  "+....+..........+.............+",
  "| id : name     : email       |",
  "+....+..........+.............+",
  "| 1  : John Doe : john@doe.dk |",
  "| 2  : Jane Doe : jane@doe.dk |",
  "| 3  : Foobar   :             |",
  "+....+..........+.............+",
]
`;

exports[`Table Table Styles default style 1`] = `
Array [
  "+----+----------+-------------+",
  "| id | name     | email       |",
  "+----+----------+-------------+",
  "| 1  | John Doe | john@doe.dk |",
  "| 2  | Jane Doe | jane@doe.dk |",
  "| 3  | Foobar   |             |",
  "+----+----------+-------------+",
]
`;

exports[`Table Table Styles slim style 1`] = `
Array [
  "┌────┬──────────┬─────────────┐",
  "│ id │ name     │ email       │",
  "├────┼──────────┼─────────────┤",
  "│ 1  │ John Doe │ john@doe.dk │",
  "│ 2  │ Jane Doe │ jane@doe.dk │",
  "│ 3  │ Foobar   │             │",
  "└────┴──────────┴─────────────┘",
]
`;

exports[`Table Table with headers 1`] = `
Array [
  "+-----+-----+",
  "| A   | B   |",
  "+-----+-----+",
  "| foo | bar |",
  "+-----+-----+",
]
`;

exports[`Table Table with no headers 1`] = `
Array [
  "+-----+-----+",
  "| a   | b   |",
  "+-----+-----+",
  "| foo | bar |",
  "+-----+-----+",
]
`;
